### PR TITLE
Added Caching to Public Competition Requests

### LINKED ISSUE #561: Cache public page + searches
Probably want to install https://github.com/chibisov/drf-extensions and make sure Redis is working on Heroku + docker-compose to cache this view

### PR DESCRIPTION
**Pull request by Logan-Ruf**
# @ mention of reviewers
@ckcollab 

# A brief description of the purpose of the changes contained in this PR.
Creates a Cache for Public Competitions Page to increase speed and lower the server load.

# Issues this PR resolves
Resolves #561 

# A checklist for hand testing
- [ ] Test Speed Difference On production
- [ ] Coche Should Update Every 15 minutes

# Checklist
- [ ] Code review by me 
- [ ] Hand tested by me 
- [ ] I'm proud of my work
- [ ] Code review by reviewer
- [ ] Hand tested by reviewer
- [ ] Ready to merge

### CODE REVIEW

File: src/apps/api/views/competitions.py [hunk 1]
```diff
@@ -36,6 +41,22 @@
 from api.permissions import IsOrganizerOrCollaborator
 
 
+class UpdatedAtKeyBit(KeyBitBase):
+    def get_data(self, **kwargs):
+        key = 'api_updated_at_timestamp'
```
Comment (ckcollab) [inline src/apps/api/views/competitions.py:46]: what is this about?

File: src/apps/api/views/competitions.py [hunk 2]
```diff
@@ -36,6 +41,22 @@
...
+        value = cache.get(key, None)
+        if not value:
+            value = datetime.datetime.utcnow()
+            cache.set(key, value=value)
+        return force_text(value)
+
+
+class CustomListKeyConstructor(DefaultKeyConstructor):
```
Comment (ckcollab) [inline src/apps/api/views/competitions.py:54]: can we import the one in drf-extensions?

File: src/settings/base.py [hunk 3]
```diff
@@ -239,6 +239,11 @@
         '%B %d, %Y',
     )
 }
+REST_FRAMEWORK_EXTENSIONS = {
+    'DEFAULT_CACHE_RESPONSE_TIMEOUT': 60 * 5,
+    'DEFAULT_OBJECT_CACHE_KEY_FUNC': 'rest_framework_extensions.utils.default_object_cache_key_func',
+    'DEFAULT_LIST_CACHE_KEY_FUNC': 'rest_framework_extensions.utils.default_list_cache_key_func',
```
Comment (ckcollab) [inline src/settings/base.py:245]: i don't think we need these?

Comment (Logan-Ruf) [reply src/settings/base.py:245]: `DEFAULT_CACHE_RESPONSE_TIMEOUT` default is none. Which means the cache would never be cleared. The other two I copied from the documentation. Might be unnecessary.

**Review by ckcollab — APPROVED** [review]
beautiful :shipit:

File: src/settings/base.py [hunk 4]
```diff
@@ -222,6 +223,30 @@
...
+                'timeout': 20,
+            },
+            'MAX_CONNECTIONS': 30,
+            'PICKLE_VERSION': -1,
+        },
+    },
+}
+REST_FRAMEWORK_EXTENSIONS = {
+    'DEFAULT_CACHE_RESPONSE_TIMEOUT': 60 * 15,
+    'DEFAULT_USE_CACHE': 'default'
```
Comment (ckcollab) [inline src/settings/base.py:247]: oh, also, do we need this?

Comment (Logan-Ruf) [reply src/settings/base.py:247]: Oh nope, it will use default automatically.
```suggestion
```

**Pull request merged**